### PR TITLE
Changes to support segregation of data loader and reader

### PR DESCRIPTION
**Pull request by hariharan-devarajan**
Changes to enable segregation of data loader and reader.

- Added support for tensorflow dataset and pytorch data loader APIs.
- Made Reader framework independent (except for TFRecord, which only works with TF)
- The pytest framework was looping within the code to do different permutations. Instead I changed it to use parameterized which allows us to filter different cases on command line. And enable running isolated tests.
- The new data loader framework (both TF and PyTorch) required the images to have same sizes. Therefore, we pad/resize the images to ensure they are same between steps. 
- The reader is responsible for reading images and batching them. Whereas the data loader converts its into framework-optimized reading.
-

### CODE REVIEW

**Comment by hariharan-devarajan** [comment]
Do opening before the dataset read for npz and csv.

**Comment by hariharan-devarajan** [comment]
Add barrier to each step and change pytest to run one test at a time.

**Comment by hariharan-devarajan** [comment]
@zhenghh04 I have made the requested changes.

File: src/reader/npz_reader.py [hunk 1]
```diff
@@ -40,48 +41,81 @@ def read(self, epoch_number):
...
-    def finalize(self):
-        pass
+    def get_sample_len(self):
+        total_samples = 0
+        for index in range(len(self._dataset)):
+            if self._dataset[index]["data"] is None:
+                with np.load(self._dataset[index]["file"], allow_pickle=True) as data:
+                    self._dataset[index]['data'] = data["x"]
+            total_samples = total_samples + self._dataset[index]['data'].shape[2]
+        return total_samples
```
Comment (zhenghh04) [inline src/reader/npz_reader.py:121]: This is too expensive; it essentially read all the dataset. If the dataset size is large, it will have trouble. Let us use the same way you used for tfreader?

File: src/reader/hdf5_reader.py [hunk 2]
```diff
@@ -63,36 +63,61 @@ def next(self):
...
+    def get_sample_len(self):
+        total_samples = 0
+        for element in self._dataset:
+            file_h5 = h5py.File(element["file"], 'r')
+            dataset = file_h5['records']
+            total_samples = total_samples + dataset.shape[0]
             file_h5.close()
-    def finalize(self):
-        pass
+        return total_samples
```
Comment (zhenghh04) [inline src/reader/hdf5_reader.py:123]: Again, let us simply use the tfreader way to calculating the total number of samples instead of open the files. Even though for HDF5, the overhead is probably small. But still, imaging all the ranks are trying to open the files.

**Review by zhenghh04 — COMMENTED** [review]
Let us discuss over this and try to resolve them today in the meeting before merging them.

File: src/reader/npz_reader.py [hunk 3]
```diff
@@ -40,48 +41,81 @@ def read(self, epoch_number):
...
+        file_index = math.floor(index / self.num_samples)
+        if self._dataset[file_index]["data"] is None:
+            with np.load(self._dataset[file_index]["file"], allow_pickle=True) as data:
+                self._dataset[file_index]['data'] = data["x"]
+        element_index = index % self.num_samples
+        my_image = self._dataset[file_index]['data'][..., element_index]
+        logging.info(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+        my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+        logging.info(f"{utcnow()} new shape of image {my_image_resized.shape}")
+        return my_image_resized
```
Comment (zhenghh04) [inline src/reader/npz_reader.py:112]: In this case again, on the second epoch, data will be read from the memory. 

It is not truly the on demand way of doing things. 

I think the key here is because you are using ```self._dataset[file_index]['data']``` to hold all the data

File: src/data_generator/jpeg_generator.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def generate(self):
                 dim1, dim2 = [max(int(d), 0) for d in random.normal(dim, dim_stdev, 2)]
             else:
                 dim1 = dim2 = dim
-            records = random.randint(255, size=(dim1, dim2, 3), dtype=np.uint8)
+            records = random.randint(255, size=(dim1, dim2), dtype=np.uint8)
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/data_generator/jpeg_generator.py:49]: We cant for two reasons.
- this will create a larger image size than described by record length.
- with the new data loader the dimensions of array should similar for all data formats. If the jpeg and png use 3d then the dataloader would not work generically.

In terms of I/O I think it is the same thing whether we use 2D or 3D till number of bytes are same for both.

File: src/data_generator/png_generator.py [hunk 5]
```diff
@@ -45,7 +45,7 @@ def generate(self):
             else:
                 dim1 = dim2 = dim
             out_path_spec = self.storage.get_uri(self._file_list[i])
-            records = random.randint(255, size=(dim1, dim2, 3), dtype=np.uint8)
+            records = random.randint(255, size=(dim1, dim2), dtype=np.uint8)
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/data_generator/png_generator.py:48]: dim is currently not taking account of channels. Also we need to keep the dimension of array same across data format for data loader to work.

File: src/data_loader/tf_data_loader.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from src.utils.utility import utcnow
+
+
+class TensorflowDataset(tf.data.Dataset):
+    def _generator(format_type, dataset_type, epoch_number):
+        format_type = format_type.decode('ascii')
+        dataset_type = dataset_type.decode('ascii')
+        logging.debug(f"{utcnow()} format_type {format_type} dataset_type {dataset_type} tensors")
+        reader = ReaderFactory.get_reader(FormatType.get_enum(format_type),
+                                          dataset_type=DatasetType.get_enum(dataset_type))
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/data_loader/tf_data_loader.py:18]: format type is data formats.
we use dataset type for train and validate.

File: src/reader/csv_reader.py [hunk 7]
```diff
@@ -39,47 +43,75 @@ def read(self, epoch_number):
...
+        file_index = math.floor(index / self.num_samples)
+        element_index = index % self.num_samples
+        if self._dataset[file_index]["data"] is None:
+            self._dataset[file_index]['data'] = pd.read_csv(self._dataset[file_index]["file"], compression="infer").to_numpy()
+        my_image = self._dataset[file_index]['data'][..., element_index]
+        logging.info(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+        my_image = np.pad(my_image, ((0, self.max_dimension - my_image.shape[0]),
+                                     (0, self.max_dimension - my_image.shape[1])),
+                          mode='constant', constant_values=0)
+        logging.info(f"{utcnow()} new shape of image {my_image.shape}")
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/reader/csv_reader.py:108]: How about in case of on demand I always open the dataframe from file and i close the dataset or remove from memory

File: src/dlio_benchmark.py [hunk 8]
```diff
@@ -210,7 +211,7 @@ def _eval(self, epoch):
             if step > total:
                 return step - 1
                 
-            self.framework.barrier()
+            #self.framework.barrier()
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/dlio_benchmark.py]: Fixed.

File: src/common/enumerations.py [hunk 9]
```diff
@@ -59,6 +59,19 @@ class DatasetType(Enum):
...
         return self.value
 
+    @staticmethod
+    def get_enum(value):
+        if DatasetType.TRAIN.value == value:
+            return DatasetType.TRAIN
+        elif DatasetType.VALID.value == value:
+            return DatasetType.VALID
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/common/enumerations.py]: Fixed

File: src/reader/csv_reader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/reader/csv_reader.py]: This makes the reading very slow.

File: src/reader/csv_reader.py [hunk 10]
```diff
@@ -39,47 +43,75 @@ def read(self, epoch_number):
...
+        return my_image
 
-    def finalize(self):
-        pass
+    def get_sample_len(self):
+        total_samples = 0
+        for index in range(len(self._dataset)):
+            if self._dataset[index]["data"] is None:
+                self._dataset[index]['data'] = pd.read_csv(self._dataset[index]["file"], compression="infer").to_numpy()
+            total_samples = total_samples + len(self._dataset[index]['data'])
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply src/reader/csv_reader.py]: Fixed

File: src/reader/npz_reader.py (hunk 1 above)
Comment (hariharan-devarajan) [reply src/reader/npz_reader.py]: Fixed

File: src/reader/hdf5_reader.py (hunk 2 above)
Comment (hariharan-devarajan) [reply src/reader/hdf5_reader.py]: fixed

File: src/reader/npz_reader.py (hunk 3 above)
Comment (hariharan-devarajan) [reply src/reader/npz_reader.py]: I can force the re-read in case of ON-DEMAND.

Comment (hariharan-devarajan) [reply src/reader/npz_reader.py]: Fixed
